opensafely/notebook.py

```diff
@@ -0,0 +1,71 @@
+import json
+from pathlib import Path
+import os
+import subprocess
+import sys
+import time
+import webbrowser
+
+
+DESCRIPTION = "Run a jupyter notebook using the OpenSAFELY environment"
+
+def add_arguments(parser):
+    pass
+
+
+def docker(*args, **kwargs):
+    cmd = ("docker",) + args
+    check = kwargs.get("check", True)
+    capture = kwargs.get("capture_output", True)
+    newlines = kwargs.get("universal_newlines", True)
+    return subprocess.run(cmd, check=check, capture_output=capture, universal_newlines=newlines)
+
+
+def main():
+    container = None
+    pwd = Path(os.getcwd())
+    name = f"opensafely-notebook-{pwd.name}"
+    launch = [
+            "run",
+            # we initially detach, so we can find out the ip address.
+            "--detach",
+            "--rm",
+            "--init",
+            # we force tty now so we can re-attach stdin/out later
+            "-it",
+            f"--name={name}",
+            f"--hostname={name}",
+            "--label=opensafely",
+            "-v",
+            f"/{pwd}:/workspace",
+            "ghcr.io/opensafely-core/python",
+            "jupyter",
+            "lab",
+            "--ip",
+            "0.0.0.0",
+            # python image currently runs as root
+            "--allow-root",
+            # disable authentication, as it's only accessible via direct access
+            # to the docker container
+            "--LabApp.token=''",
+        ]
+
+    launch_ps = docker(*launch)
+    container = launch_ps.stdout.strip()
+    # let jupyter get started
+    time.sleep(1)
+
+    # figure out the url
+    inspect_ps = docker("inspect", container)
+    metadata = json.loads(inspect_ps.stdout)[0]
+    ip = metadata["NetworkSettings"]["IPAddress"]
+    url = f"http://{ip}:8888"
+
+    # open a webbrowser pointing to the docker container
+    webbrowser.open(url, new=2)
+
+    # show what the server has said so far
+    docker("logs", container, capture_output=False)
+
+    # take over the shell so we see further info from the jupyter server
+    docker("container", "attach", container, capture_output=False)
```

Comment:
I did [something](https://github.com/opensafely/server-instructions/blob/master/scripts/database-notebook.sh) sort of similar to this for running notebooks on the TPP server with database access. And I seem to remember that shutting down Jupyter via the web interface still leaves the container running on Windows, although it terminates it on Linux.

It might be worth waiting (or just looping on sleep) for the process in Python and then explicitly terminating the container on exit.

Reply:
Ah, I haven't tried that, will try this whole thing on windows

Reply:
In linux, it does shutdown. About to try in on windows.